### PR TITLE
 Zwave: implement CLOCK and TIME_PARAMETERS command class

### PR DESCRIPTION
**Pull request by jongj**
The clock on the node is requested during DYNAMIC_VALUES stage.
If its not set correctly the time will be updated. 
The calculated time difference, time offset is reported in a Event.

### CODE REVIEW

File: addons/binding/org.openhab.binding.zwave/src/main/java/org/openhab/binding/zwave/internal/protocol/commandclass/ZWaveClockCommandClass.java [hunk 1]
```diff
@@ -34,6 +48,13 @@
     private static final int CLOCK_GET = 5;
     private static final int CLOCK_REPORT = 6;
 
+    // acceptable amount of time the clock on the node may be off before
+    // we provide the node with a clock update.
+    private static final int CLOCK_PRESISION = 5 * 60 * 1000;
```
Comment (cdjackson) [inline addons/binding/org.openhab.binding.zwave/src/main/java/org/openhab/binding/zwave/internal/protocol/commandclass/ZWaveClockCommandClass.java]: Minor spelling mistake: CLOCK_PRECISION
:)

File: addons/binding/org.openhab.binding.zwave/src/main/java/org/openhab/binding/zwave/internal/protocol/commandclass/ZWaveClockCommandClass.java [hunk 2]
```diff
@@ -64,10 +134,70 @@ public void handleApplicationCommandRequest(SerialMessage serialMessage, int off
...
+                int javaDay = day == 7 ? 1 : day + 1;
+                logger.debug(String.format("NODE %d: Received clock report: %s %02d:%02d", getNode().getNodeId(),
+                        days[javaDay], hour, minute));
+
+                Calendar cal = Calendar.getInstance();
+                cal.set(Calendar.DAY_OF_WEEK, javaDay);
+                cal.set(Calendar.HOUR_OF_DAY, hour);
+                cal.set(Calendar.MINUTE, minute);
+                Date nodeTime = cal.getTime();
+                long clockOffset = (nodeTime.getTime() - System.currentTimeMillis()) / 1000 / 60;
```
Comment (cdjackson) [inline addons/binding/org.openhab.binding.zwave/src/main/java/org/openhab/binding/zwave/internal/protocol/commandclass/ZWaveClockCommandClass.java]: I wonder if this is the best approach for the protocol layers to make this decision. I think it's better to simply report the time and let the application layers calculate the offset and decide what to do about it.

File: addons/binding/org.openhab.binding.zwave/src/main/java/org/openhab/binding/zwave/internal/protocol/commandclass/ZWaveClockCommandClass.java [hunk 3]
```diff
@@ -64,10 +134,70 @@ public void handleApplicationCommandRequest(SerialMessage serialMessage, int off
...
+            logger.warn("NODE {}: adjusting clock", getNode().getNodeId());
+            updateClock = false;
+            result.add(getSetMessage(new Date()));
+            result.add(getValueMessage());
+        }
+        return result;
+    }
+
+    /**
+     * Z-Wave Clock Event class. Indicates the current clock offset in minutes.
```
Comment (cdjackson) [inline addons/binding/org.openhab.binding.zwave/src/main/java/org/openhab/binding/zwave/internal/protocol/commandclass/ZWaveClockCommandClass.java]: Again, I think it's better to simply provide the time - not the offset?

File: addons/binding/org.openhab.binding.zwave/src/main/java/org/openhab/binding/zwave/internal/protocol/commandclass/ZWaveClockCommandClass.java [hunk 4]
```diff
@@ -64,10 +134,70 @@ public void handleApplicationCommandRequest(SerialMessage serialMessage, int off
...
         }
     }
+
+    @Override
+    public Collection<SerialMessage> getDynamicValues(boolean refresh) {
+        ArrayList<SerialMessage> result = new ArrayList<SerialMessage>();
+        if (refresh == true && getEndpoint() == null) {
+            result.add(getValueMessage());
+        }
+        if (updateClock) {
```
Comment (cdjackson) [inline addons/binding/org.openhab.binding.zwave/src/main/java/org/openhab/binding/zwave/internal/protocol/commandclass/ZWaveClockCommandClass.java]: I think I'd again be tempted to remove the updating of the time and do it in the application. It's more flexible if it's done there.

File: addons/binding/org.openhab.binding.zwave/src/main/java/org/openhab/binding/zwave/internal/protocol/commandclass/ZWaveTimeParametersCommandClass.java [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+     * @throws ZWaveSerialMessageException
+     */
+    @Override
+    public void handleApplicationCommandRequest(SerialMessage serialMessage, int offset, int endpoint)
+            throws ZWaveSerialMessageException {
+        logger.debug("NODE {}: Received clock command (v{})", this.getNode().getNodeId(), this.getVersion());
+        int command = serialMessage.getMessagePayloadByte(offset);
+        switch (command) {
+            case TIME_GET:
+                logger.debug("NODE %d: requested the time", getNode().getNodeId());
```
Comment (cdjackson) [inline addons/binding/org.openhab.binding.zwave/src/main/java/org/openhab/binding/zwave/internal/protocol/commandclass/ZWaveTimeParametersCommandClass.java]: The debug message is incorrecly formatted - you should use {}

File: addons/binding/org.openhab.binding.zwave/src/main/java/org/openhab/binding/zwave/internal/protocol/commandclass/ZWaveTimeParametersCommandClass.java [hunk 6]
```diff
@@ -0,0 +1,208 @@
...
+                cal.clear();
+                cal.set(year, month - 1, day, hour, minute, second);
+                logger.debug("NODE {}: Received time report: {}", getNode().getNodeId(), cal.getTime());
+
+                Date nodeTime = cal.getTime();
+                long clockOffset = (nodeTime.getTime() - System.currentTimeMillis()) / 1000;
+                ZWaveTimeValueEvent zEvent = new ZWaveTimeValueEvent(getNode().getNodeId(), endpoint, clockOffset);
+                this.getController().notifyEventListeners(zEvent);
+
+                Date low = new Date(System.currentTimeMillis() - CLOCK_PRESISION);
```
Comment (cdjackson) [inline addons/binding/org.openhab.binding.zwave/src/main/java/org/openhab/binding/zwave/internal/protocol/commandclass/ZWaveTimeParametersCommandClass.java]: Same comments as the CLOCK class - I'd suggest to return the time, not the offset.

File: addons/binding/org.openhab.binding.zwave/src/main/java/org/openhab/binding/zwave/internal/protocol/commandclass/ZWaveTimeParametersCommandClass.java [hunk 7]
```diff
@@ -0,0 +1,208 @@
...
+                Date high = new Date(System.currentTimeMillis() + CLOCK_PRESISION);
+
+                // auto correct the time if off by more than 1 minutes
+                if (nodeTime.before(low) || nodeTime.after(high)) {
+                    updateClock = true;
```
Comment (cdjackson) [inline addons/binding/org.openhab.binding.zwave/src/main/java/org/openhab/binding/zwave/internal/protocol/commandclass/ZWaveTimeParametersCommandClass.java]: Same again - I'd suggest to remove this and do it in the application where we have more flexibility

File: addons/binding/org.openhab.binding.zwave/src/main/java/org/openhab/binding/zwave/internal/protocol/commandclass/ZWaveTimeParametersCommandClass.java [hunk 8]
```diff
@@ -0,0 +1,208 @@
...
+        }
+    }
+
+    @Override
+    public Collection<SerialMessage> getDynamicValues(boolean refresh) {
+        ArrayList<SerialMessage> result = new ArrayList<SerialMessage>();
+        if (refresh == true && getEndpoint() == null) {
+            result.add(getValueMessage());
+        }
+        if (updateClock) {
```
Comment (cdjackson) [inline addons/binding/org.openhab.binding.zwave/src/main/java/org/openhab/binding/zwave/internal/protocol/commandclass/ZWaveTimeParametersCommandClass.java]: Again - I'd suggest to remove the setting of the time from the command class.

File: addons/binding/org.openhab.binding.zwave/src/main/java/org/openhab/binding/zwave/internal/protocol/commandclass/ZWaveTimeParametersCommandClass.java [hunk 9]
```diff
@@ -0,0 +1,208 @@
...
+     */
+    @Override
+    public void handleApplicationCommandRequest(SerialMessage serialMessage, int offset, int endpoint)
+            throws ZWaveSerialMessageException {
+        logger.debug("NODE {}: Received clock command (v{})", this.getNode().getNodeId(), this.getVersion());
+        int command = serialMessage.getMessagePayloadByte(offset);
+        switch (command) {
+            case TIME_GET:
+                logger.debug("NODE %d: requested the time", getNode().getNodeId());
+                getController().sendData(getSetMessage(new Date()));
```
Comment (cdjackson) [inline addons/binding/org.openhab.binding.zwave/src/main/java/org/openhab/binding/zwave/internal/protocol/commandclass/ZWaveTimeParametersCommandClass.java]: I'm not a fan (in general) of calling back to the controller. I know it's done quite a lot in the binding (at the moment), but I'd quite like to reduce or eliminate it where possible...

Additionally, I think we should move all the 'time set' functionallity out of the command classes and into the application where there is more flexibility and knowledge of the environment.

**Comment by cdjackson** [comment]
Thanks for this - it's something I've been meaning to look at for a while :)

My main comment is that I would remove all the time set autonomy from the command classes and we should instead put it in the thing handler - using an event to pass the request(s). This allows us to be a bit smarter - we might only want to set the time on a node at 4am for example, so if a node requests the time, we might want to check how far off it is, and what the current time is, and only set it then - or whatever...

Also, if we move the 'precision' checks out of the command class, we can at some stage make this time configurable - that's more difficult in the command class.

I'm not suggesting to add any of the above right now - just to move the time setting out of the command classes so that it allows us to add functionality later. I would prefer to try and keep the command classes 'dumb' and let the application do more.

Let me know if that doesn't make sense, or you disagree and we can discuss more :)

**Comment by jongj** [comment]
I agree that we need more control and that the CC's are not the place for that.  So I removed the auto time update 'feature', the evens now report the date and not the offset. 

What remains is to update the application to respond to the events. Do you want me to give it a try?
and add the event processing to ZWaveThingHandler.ZWaveIncomingEvent? I also don't mind if you merge this PR as it is now. 

The nodes that I have that use the CLOCK CC are very bad at keeping time, there clock  drifts minutes per day. So correcting the time during startup is not sufficient, but that's maybe work for a follow up PR.

**Comment by cdjackson** [comment]
Ok - thanks. I’ll take a look over this again shortly and will likely merge. I’ll then add something into the handler to handle this. It would be easy to poll the time, and if the time is out, to reset it, so it’s not just done on startup.

**Comment by cdjackson** [comment]
Thanks - looks good.

**Comment by cdjackson** [comment]
@jongj I have implemented a new channel type "time_offset". Can you update at least one or two of your devices to add this channel to the database for the CLOCK and TIME_PARAMETERS class (or tell me what the device is and I'll do the first one).

**Comment by jongj** [comment]
CLOCK is used by http://www.cd-jackson.com/index.php/zwave/zwave-device-database/zwave-device-list/devicesummary/165 

The device that uses TIME_PARAMETERS is not in the database yet. waiting on #812

**Comment by jongj** [comment]
I updated http://www.cd-jackson.com/index.php/zwave/zwave-device-database/zwave-device-list/devicesummary/129 with a time_paramter, but export does not show the new channel or do I need to wait for you to approve the update before I can see it?

**Comment by cdjackson** [comment]
I’ve only implemented the export for the CLOCK so far - give me a couple of minutes and I’ll add the TIME_PARAMETERS

**Comment by cdjackson** [comment]
Done…

**Comment by jongj** [comment]
ok, the channel show up now :smile:

**Comment by cdjackson** [comment]
@jongj please take a look at my zwave_master branch and see if it is working. I've not tested on the CLOCK classes.

What is implemented is the time_offset channel. When this channel is updated, it will perform a check to see if the time is off, and if so it will update (if everything works!). In theory, you should be able to configure the allowable offset in the channel configuration - it allows setting from 5 seconds to 20 minutes I think.

I'll also load a new HABmin JAR to the HABmin repo shortly. This has a REFRESH button in the thing overview area so it will allow you to poll the device as you like.

**Comment by cdjackson** [comment]
HABmin JAR now also updated in the /output folder if you want to use it...

**Pull request merged**